File: autocontext/src/autocontext/integrations/anthropic/_proxy.py
"""ClientProxy — attribute-delegating wrapper around Anthropic clients."""
from __future__ import annotations

import time
import traceback
from datetime import UTC, datetime
from typing import Any

from ulid import ULID

from autocontext.integrations._shared.session import current_session
from autocontext.integrations._shared.sink import TraceSink
from autocontext.integrations.anthropic._taxonomy import map_exception_to_reason
from autocontext.integrations.anthropic._trace_builder import (
    build_failure_trace,
    build_request_snapshot,
    build_success_trace,
    finalize_streaming_trace,
)
from autocontext.production_traces.hashing import (
    hash_session_id,
    hash_user_id,
    load_install_salt,
)

_WRAPPED_SENTINEL = "__autocontext_wrapped__"


def _now_iso() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def _is_async_client(client: Any) -> bool:
    try:
        from anthropic import AsyncAnthropic  # noqa: PLC0415
        return isinstance(client, AsyncAnthropic)
    except ImportError:
        pass
    return type(client).__name__.startswith("Async")


def _resolve_identity(per_call: dict[str, Any] | None) -> dict[str, str]:
    raw: dict[str, str] = {}
    if per_call:
        if "user_id" in per_call and per_call["user_id"] is not None:
            raw["user_id"] = per_call["user_id"]
        if "session_id" in per_call and per_call["session_id"] is not None:
            raw["session_id"] = per_call["session_id"]
    if not raw:
        raw = dict(current_session())
    if not raw:
        return {}
    salt = load_install_salt(".") or ""
    hashed: dict[str, str] = {}
    if "user_id" in raw:
        hashed["user_id_hash"] = hash_user_id(raw["user_id"], salt)
    if "session_id" in raw:
        hashed["session_id_hash"] = hash_session_id(raw["session_id"], salt)
    return hashed


def _response_usage_and_content(response: Any) -> tuple[dict[str, Any] | None, list[dict[str, Any]], str | None]:
    usage = None
    if getattr(response, "usage", None):
        usage = response.usage.model_dump() if hasattr(response.usage, "model_dump") else dict(response.usage)
    content = response.content if hasattr(response, "content") else []
    stop_reason = response.stop_reason if hasattr(response, "stop_reason") else None
    if content and not isinstance(content[0], dict):
        content_list = [b.model_dump() if hasattr(b, "model_dump") else dict(b) for b in content]
    else:
        content_list = list(content)
    return usage, content_list, stop_reason


class _MessagesProxy:
    def __init__(self, parent: ClientProxy) -> None:
        self._parent = parent

    def create(self, **kwargs: Any) -> Any:
        stream = kwargs.get("stream", False)
        if stream:
            if self._parent._is_async:
                return self._parent._invoke_streaming_async(kwargs=kwargs)
            return self._parent._invoke_streaming(kwargs=kwargs)
        if self._parent._is_async:
            return self._parent._invoke_non_streaming_async(kwargs=kwargs)
        return self._parent._invoke_non_streaming(kwargs=kwargs)

    def stream(self, **kwargs: Any) -> Any:
        if self._parent._is_async:
            return self._parent._invoke_helper_streaming_async(kwargs=kwargs)
        return self._parent._invoke_helper_streaming(kwargs=kwargs)


class ClientProxy:
    def __init__(
        self,
        *,
        inner: Any,
        sink: TraceSink,
        app_id: str,
        environment_tag: str,
    ) -> None:
        object.__setattr__(self, "_inner", inner)
        object.__setattr__(self, "_sink", sink)
        object.__setattr__(self, "_app_id", app_id)
        object.__setattr__(self, "_environment_tag", environment_tag)
        object.__setattr__(self, "_is_async", _is_async_client(inner))
        object.__setattr__(self, _WRAPPED_SENTINEL, True)

    def __getattr__(self, name: str) -> Any:
        if name == "messages":
            return _MessagesProxy(self)
        return getattr(self._inner, name)

    def _source_info(self) -> dict[str, Any]:
        try:
            from importlib.metadata import version
            ver = version("autocontext")
        except Exception:
            ver = "0.0.0"
        return {"emitter": "sdk", "sdk": {"name": "autocontext-py", "version": ver}}

    def _env(self) -> dict[str, Any]:
        return {"environmentTag": self._environment_tag, "appId": self._app_id}

    def _invoke_non_streaming(self, *, kwargs: dict[str, Any]) -> Any:
        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = self._inner.messages.create(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage, content_list, stop_reason = _response_usage_and_content(response)
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_content=content_list,
            response_usage=usage,
            response_stop_reason=stop_reason,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    async def _invoke_non_streaming_async(self, *, kwargs: dict[str, Any]) -> Any:
        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        try:
            response = await self._inner.messages.create(**kwargs)
        except Exception as exc:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=self._env(),
                source_info=self._source_info(),
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            self._sink.add(trace)
            raise
        ended_at = _now_iso()
        latency_ms = int((time.monotonic() - started_monotonic) * 1000)
        usage, content_list, stop_reason = _response_usage_and_content(response)
        trace = build_success_trace(
            request_snapshot=request_snapshot,
            response_content=content_list,
            response_usage=usage,
            response_stop_reason=stop_reason,
            identity=identity,
            timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
            env=self._env(),
            source_info=self._source_info(),
            trace_id=str(ULID()),
        )
        self._sink.add(trace)
        return response

    def _invoke_streaming(self, *, kwargs: dict[str, Any]) -> Any:
        from autocontext.integrations.anthropic._stream import StreamProxy  # noqa: PLC0415
        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        inner_stream = self._inner.messages.create(**kwargs)
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_finalize(acc: Any, outcome: dict[str, Any]) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = finalize_streaming_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                accumulated_content_blocks=acc.content_blocks,
                accumulated_usage=acc.usage or None,
                accumulated_stop_reason=acc.stop_reason,
                outcome=outcome,
            )
            sink.add(trace)

        return StreamProxy(inner_stream=inner_stream, on_finalize=on_finalize)

    def _invoke_helper_streaming(self, *, kwargs: dict[str, Any]) -> Any:
        from autocontext.integrations.anthropic._stream import HelperStreamManagerProxy  # noqa: PLC0415

        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_success(message: Any) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            usage, content_list, stop_reason = _response_usage_and_content(message)
            trace = build_success_trace(
                request_snapshot=request_snapshot,
                response_content=content_list,
                response_usage=usage,
                response_stop_reason=stop_reason,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
            )
            sink.add(trace)

        def on_failure(exc: BaseException) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            sink.add(trace)

        def on_partial(message: Any) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            usage, content_list, stop_reason = _response_usage_and_content(message)
            trace = build_success_trace(
                request_snapshot=request_snapshot,
                response_content=content_list,
                response_usage=usage,
                response_stop_reason=stop_reason,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
            )
            trace["outcome"] = {"label": "partial", "reasoning": "abandonedStream"}
            sink.add(trace)

        inner_manager = self._inner.messages.stream(**kwargs)
        return HelperStreamManagerProxy(
            inner_manager=inner_manager,
            on_success=on_success,
            on_failure=on_failure,
            on_partial=on_partial,
        )

    def _invoke_streaming_async(self, *, kwargs: dict[str, Any]) -> Any:
        import inspect  # noqa: PLC0415

        from autocontext.integrations.anthropic._stream import AsyncStreamProxy  # noqa: PLC0415
        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_finalize(acc: Any, outcome: dict[str, Any]) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = finalize_streaming_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                accumulated_content_blocks=acc.content_blocks,
                accumulated_usage=acc.usage or None,
                accumulated_stop_reason=acc.stop_reason,
                outcome=outcome,
            )
            sink.add(trace)

        async def _make_proxy() -> AsyncStreamProxy:
            raw = self._inner.messages.create(**kwargs)
            if inspect.iscoroutine(raw) or hasattr(raw, "__await__"):
                inner_stream = await raw
            else:
                inner_stream = raw
            return AsyncStreamProxy(inner_stream=inner_stream, on_finalize=on_finalize)

        return _make_proxy()

    def _invoke_helper_streaming_async(self, *, kwargs: dict[str, Any]) -> Any:
        from autocontext.integrations.anthropic._stream import AsyncHelperStreamManagerProxy  # noqa: PLC0415

        per_call = kwargs.pop("autocontext", None)
        identity = _resolve_identity(per_call)
        request_snapshot = build_request_snapshot(
            model=kwargs.get("model", ""),
            messages=kwargs.get("messages", []),
            extra_kwargs={k: v for k, v in kwargs.items() if k not in {"model", "messages"}},
        )
        started_at = _now_iso()
        started_monotonic = time.monotonic()
        sink = self._sink
        env = self._env()
        source_info = self._source_info()

        def on_success(message: Any) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            usage, content_list, stop_reason = _response_usage_and_content(message)
            trace = build_success_trace(
                request_snapshot=request_snapshot,
                response_content=content_list,
                response_usage=usage,
                response_stop_reason=stop_reason,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
            )
            sink.add(trace)

        def on_failure(exc: BaseException) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            trace = build_failure_trace(
                request_snapshot=request_snapshot,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
                reason_key=map_exception_to_reason(exc),
                error_message=str(exc),
                stack=traceback.format_exc(),
            )
            sink.add(trace)

        def on_partial(message: Any) -> None:
            ended_at = _now_iso()
            latency_ms = int((time.monotonic() - started_monotonic) * 1000)
            usage, content_list, stop_reason = _response_usage_and_content(message)
            trace = build_success_trace(
                request_snapshot=request_snapshot,
                response_content=content_list,
                response_usage=usage,
                response_stop_reason=stop_reason,
                identity=identity,
                timing={"startedAt": started_at, "endedAt": ended_at, "latencyMs": latency_ms},
                env=env,
                source_info=source_info,
                trace_id=str(ULID()),
            )
            trace["outcome"] = {"label": "partial", "reasoning": "abandonedStream"}
            sink.add(trace)

        inner_manager = self._inner.messages.stream(**kwargs)
        return AsyncHelperStreamManagerProxy(
            inner_manager=inner_manager,
            on_success=on_success,
            on_failure=on_failure,
            on_partial=on_partial,
        )

File: ts/src/integrations/anthropic/proxy.ts
/**
 * ClientProxy — Proxy-based wrapper around an Anthropic client.
 *
 * Intercepts .messages.create and .messages.stream. All other attribute
 * access passes through transparently. Mirror of Python _proxy.py for Anthropic.
 */
import { ulid } from "ulid";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import type { TraceSink } from "../_shared/sink.js";
import { currentSession } from "../_shared/session.js";
import { mapExceptionToReason } from "./taxonomy.js";
import {
  buildRequestSnapshot,
  buildSuccessTrace,
  buildFailureTrace,
  finalizeStreamingTrace,
  type RequestSnapshot,
} from "./trace-builder.js";
import { AnthropicStreamProxy, wrapHelperStream } from "./stream-proxy.js";
import {
  hashUserId,
  hashSessionId,
  installSaltPath,
} from "../../production-traces/sdk/hashing.js";
import type { ContentBlock } from "./content.js";

export const WRAPPED_SENTINEL = Symbol.for("autocontext.wrapped");

function _nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function _resolveIdentity(
  perCall: Record<string, string> | null | undefined,
): Record<string, string> {
  let raw: Record<string, string> = {};
  if (perCall) {
    if (perCall["user_id"] != null) raw["user_id"] = perCall["user_id"];
    if (perCall["session_id"] != null) raw["session_id"] = perCall["session_id"];
  }
  if (Object.keys(raw).length === 0) {
    const ambient = currentSession();
    if (ambient.userId) raw["user_id"] = ambient.userId;
    if (ambient.sessionId) raw["session_id"] = ambient.sessionId;
  }
  if (Object.keys(raw).length === 0) return {};
  const salt = _loadSaltSync(".");
  if (!salt) return {};
  const hashed: Record<string, string> = {};
  if (raw["user_id"]) hashed["user_id_hash"] = hashUserId(raw["user_id"], salt);
  if (raw["session_id"]) hashed["session_id_hash"] = hashSessionId(raw["session_id"], salt);
  return hashed;
}

function _loadSaltSync(cwd: string): string | null {
  try {
    const saltPath = installSaltPath(cwd);
    if (!existsSync(saltPath)) return null;
    const content = readFileSync(saltPath, "utf-8").trim();
    return content || null;
  } catch {
    return null;
  }
}

let _cachedVersion: string | null = null;

function _resolvePackageVersion(): string {
  if (_cachedVersion !== null) return _cachedVersion;
  try {
    let dir = dirname(fileURLToPath(import.meta.url));
    for (let depth = 0; depth < 10; depth++) {
      const candidate = join(dir, "package.json");
      if (existsSync(candidate)) {
        const pkg = JSON.parse(readFileSync(candidate, "utf-8")) as {
          name?: string;
          version?: string;
        };
        if (pkg.name === "autoctx" && typeof pkg.version === "string") {
          _cachedVersion = pkg.version;
          return _cachedVersion;
        }
      }
      const parent = dirname(dir);
      if (parent === dir) break;
      dir = parent;
    }
  } catch {
    // best-effort
  }
  _cachedVersion = "0.0.0";
  return _cachedVersion;
}

function _responseUsageAndContent(resp: Record<string, unknown>): {
  usage: Record<string, unknown> | null;
  content: ContentBlock[];
  stopReason: string | null;
} {
  return {
    usage: (resp["usage"] as Record<string, unknown>) ?? null,
    content: ((resp["content"] as Array<Record<string, unknown>>) ?? []) as ContentBlock[],
    stopReason: (resp["stop_reason"] as string) ?? null,
  };
}

export class ClientProxy {
  readonly _inner: unknown;
  readonly _sink: TraceSink;
  readonly _appId: string;
  readonly _environmentTag: string;

  constructor(opts: {
    inner: unknown;
    sink: TraceSink;
    appId: string;
    environmentTag: string;
  }) {
    this._inner = opts.inner;
    this._sink = opts.sink;
    this._appId = opts.appId;
    this._environmentTag = opts.environmentTag;
  }

  _sourceInfo(): { emitter: string; sdk: { name: string; version: string } } {
    return {
      emitter: "sdk",
      sdk: { name: "autocontext-ts", version: _resolvePackageVersion() },
    };
  }

  _env(): { environmentTag: string; appId: string } {
    return { environmentTag: this._environmentTag, appId: this._appId };
  }

  async _invokeNonStreaming(kwargs: Record<string, unknown>): Promise<unknown> {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = _resolveIdentity(perCall);
    const snapshot: RequestSnapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const startedAt = _nowIso();
    const startedMonotonic = Date.now();
    let resp: unknown;
    try {
      const inner = this._inner as {
        messages: { create: (k: unknown) => Promise<unknown> };
      };
      resp = await inner.messages.create(kwargs);
    } catch (exc) {
      const endedAt = _nowIso();
      const latencyMs = Date.now() - startedMonotonic;
      const trace = buildFailureTrace({
        requestSnapshot: snapshot,
        identity,
        timing: { startedAt, endedAt, latencyMs },
        env: this._env(),
        sourceInfo: this._sourceInfo(),
        traceId: ulid(),
        reasonKey: mapExceptionToReason(exc),
        errorMessage: String(exc),
        stack: exc instanceof Error ? (exc.stack ?? null) : null,
      });
      this._sink.add(trace as unknown as Record<string, unknown>);
      throw exc;
    }
    const endedAt = _nowIso();
    const latencyMs = Date.now() - startedMonotonic;
    const r = resp as Record<string, unknown>;
    const usage = (r["usage"] as Record<string, unknown>) ?? null;
    const content = (r["content"] as Array<Record<string, unknown>>) ?? [];
    const stopReason = (r["stop_reason"] as string) ?? null;
    const trace = buildSuccessTrace({
      requestSnapshot: snapshot,
      responseContent: content as unknown as ContentBlock[],
      responseUsage: usage,
      responseStopReason: stopReason,
      identity,
      timing: { startedAt, endedAt, latencyMs },
      env: this._env(),
      sourceInfo: this._sourceInfo(),
      traceId: ulid(),
    });
    this._sink.add(trace as unknown as Record<string, unknown>);
    return resp;
  }

  _invokeStreaming(kwargs: Record<string, unknown>): AnthropicStreamProxy {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = _resolveIdentity(perCall);
    const snapshot: RequestSnapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const startedAt = _nowIso();
    const startedMonotonic = Date.now();
    const inner = this._inner as {
      messages: { create: (k: unknown) => unknown };
    };
    const rawStream = inner.messages.create(kwargs);
    const sink = this._sink;
    const env = this._env();
    const sourceInfo = this._sourceInfo();

    const onFinalize = (
      blocks: Map<number, import("./trace-builder.js").AccumulatedBlock>,
      usage: Record<string, unknown> | null,
      stopReason: string | null,
      outcome: Record<string, unknown>,
    ): void => {
      const endedAt = _nowIso();
      const latencyMs = Date.now() - startedMonotonic;
      const trace = finalizeStreamingTrace({
        requestSnapshot: snapshot,
        identity,
        timing: { startedAt, endedAt, latencyMs },
        env,
        sourceInfo,
        traceId: ulid(),
        accumulatedContentBlocks: blocks,
        accumulatedUsage: usage,
        accumulatedStopReason: stopReason,
        outcome,
      });
      sink.add(trace as unknown as Record<string, unknown>);
    };

    return new AnthropicStreamProxy({ innerStream: rawStream, onFinalize });
  }

  _invokeHelperStreaming(kwargs: Record<string, unknown>): unknown {
    const perCall = kwargs["autocontext"] as Record<string, string> | null;
    delete kwargs["autocontext"];
    const identity = _resolveIdentity(perCall);
    const snapshot: RequestSnapshot = buildRequestSnapshot({
      model: String(kwargs["model"] ?? ""),
      messages: (kwargs["messages"] as Array<Record<string, unknown>>) ?? [],
      extraKwargs: Object.fromEntries(
        Object.entries(kwargs).filter(([k]) => k !== "model" && k !== "messages"),
      ),
    });
    const startedAt = _nowIso();
    const startedMonotonic = Date.now();
    const inner = this._inner as {
      messages: { stream: (k: unknown) => unknown };
    };
    const helper = inner.messages.stream(kwargs);
    const sink = this._sink;
    const env = this._env();
    const sourceInfo = this._sourceInfo();

    return wrapHelperStream({
      innerHelper: helper,
      onFinalize: (message, outcome) => {
        const endedAt = _nowIso();
        const latencyMs = Date.now() - startedMonotonic;
        const { usage, content, stopReason } = _responseUsageAndContent(message);
        const trace = buildSuccessTrace({
          requestSnapshot: snapshot,
          responseContent: content,
          responseUsage: usage,
          responseStopReason: stopReason,
          identity,
          timing: { startedAt, endedAt, latencyMs },
          env,
          sourceInfo,
          traceId: ulid(),
        });
        sink.add(
          { ...trace, outcome: outcome as typeof trace.outcome } as unknown as Record<string, unknown>,
        );
      },
      onFailure: (exc) => {
        const endedAt = _nowIso();
        const latencyMs = Date.now() - startedMonotonic;
        const trace = buildFailureTrace({
          requestSnapshot: snapshot,
          identity,
          timing: { startedAt, endedAt, latencyMs },
          env,
          sourceInfo,
          traceId: ulid(),
          reasonKey: mapExceptionToReason(exc),
          errorMessage: String(exc),
          stack: exc instanceof Error ? (exc.stack ?? null) : null,
        });
        sink.add(trace as unknown as Record<string, unknown>);
      },
    });
  }
}
